### PR TITLE
Bug/1335 gmap points

### LINKED ISSUE #1335: GMap points don't move consistently
When I move around the GMap Plot the points don't move consistently move with the map and appear in a different location.

screen shot 2014-10-23 at 4 34 07 pm
screen shot 2014-10-23 at 4 33 29 pm

We are using this script (plotting.py):

```python
from bokeh.embed import components
from bokeh.glyphs import Circle
from bokeh.objects import (
    GMapPlot, Range1d, ColumnDataSource,
    PanTool, WheelZoomTool, BoxSelectTool,
    BoxSelectionOverlay, GMapOptions)
from bokeh.resources import CDN

from bokeh.plotting import scatter

def map_builder(column_data_source):

    x_range = Range1d()
    y_range = Range1d()
 
    map_options = GMapOptions(lat=10, lng=-8, zoom=2) # centers on West Africa

    plot = GMapPlot(
        x_range=x_range, y_range=y_range,
        map_options=map_options,
        # title = "Map"  # Should probably pass a name to map function
    )

    plot.map_options.map_type="hybrid"
    circle = Circle(x="longitude", y="latitude", size=10, fill_color="fill", line_color="black")
    plot.add_glyph(column_data_source, circle)

    pan = PanTool()
    wheel_zoom = WheelZoomTool()
    # TODO Add hover tooltip with extra column data
    plot.add_tools(pan, wheel_zoom)

    return components(plot, CDN)
```

A complete example would be:

```python
from blaze import *
from bokeh.objects import ColumnDataSource
from plotting import map_builder
 t = Data('/data/geo.csv')
t2 = t[['longitude', 'latitude']]
cds = into(ColumnDataSource, t2)
script, div = map_builder(cds)
```

I'll send you the file.

cc: @bryevdv

### PR DESCRIPTION
**Pull request by birdsarah**
Fixes #1335 

1) Improves gmap interaction code
2) Projects xy coordinates into web mercator so they line up with tile layer

Could be split into two PRs for each of the above components.

Does not have tests. Not sure how to write tests for (1). Could probably write tests for (2).

### CODE REVIEW

File: bokehjs/package.json [hunk 1]
```diff
@@ -77,6 +77,7 @@
     "gear_utils": "./src/vendor/gear-utils/gear-utils.js",
     "kiwi": "./src/vendor/kiwi/kiwi.js",
     "gloo2": "./src/vendor/gloo/gloo2.js",
+    "proj4": "/Users/caged/Dev/bokeh/bokeh/bokehjs/src/vendor/proj4/dist/proj4.js",
```
Comment (mattpap) [inline bokehjs/package.json]: Use a relative path.

Comment (birdsarah) [inline bokehjs/package.json]: doh! thanks :D

**Comment by mattpap** [comment]
@birdsarah, please remove all unnecessary files from proj4's directory. This includes build files, .npm-whatever, .jshints-whatever and so on. There should be only unminified JS/CSS and images left. See other libraries under `vendor` for reference. In general, we don't want to carry extra garbage in the repository.

**Comment by birdsarah** [comment]
@mattpap, fair enough. Before I remove everything I'd like further feedback:
* proj4 can be npm installed - http://proj4js.org/
* I built it locally because when you build you have the option to specify the projections that you want, and therefore you don't build more than you need to
* If we just put the built, minified js in the vendor directory, what should we leave by way of documentation so that people can rebuild when they need to / know how what was there was built - just a README?

**Comment by mattpap** [comment]
Which is the generated file? I presume `dist/proj4.js`. What's the difference in size between npm installed version and the generated one? I presume it's big enough. In such a case, we should include a non-minified version of the custom build version and include some dev documentation. I'm not sure if a README in `src/vendor/proj4` is the right location. I think we should have some place in the actual
documentation for this (as we should have for `bk-` prefixing, etc.).

**Comment by mattpap** [comment]
One thing to consider is whether we have to have a custom built version at all. Maybe it all boils down to just managing `require()`s in some bokehjs' module. Including a custom built version is fine, but if we can avoid extra build step, then it would be awesome.

**Comment by mattpap** [comment]
`src/vendor/proj4` should be renamed to `src/vendor/proj4-version-number` (presumably 2.3.10, that's the latest tag on github).

**Comment by mattpap** [comment]
Now I see this is 2.3.11-alpha. You should have used a stable tag.

**Comment by birdsarah** [comment]
ok, i've pushed the npm install version. The difference in size is as follows:

### Before addition of proj4
Build artifact sizes:
- bokeh.js      : 2304.0 KB
- bokeh.css     :  555.0 KB
- bokeh.min.js  : 1111.0 KB
- bokeh.min.css :  526.0 KB

### With custom build
Build artifact sizes:
 - bokeh.js      : 2336.0 KB
 - bokeh.css     :  555.0 KB
 - bokeh.min.js  : 1142.0 KB
 - bokeh.min.css :  526.0 KB

### With npm install
Build artifact sizes
 - bokeh.js      : 2461.0 KB
 - bokeh.css     :  555.0 KB
 - bokeh.min.js  : 1185.0 KB
 - bokeh.min.css :  526.0 KB

**Comment by birdsarah** [comment]
am just looking into the fact that the webgl maps_cities.py is not properly aligned (its not worse than it was before)

**Comment by birdsarah** [comment]
ping @almarklein

**Comment by birdsarah** [comment]
Good news, there appears to be nothing wrong with the webgl - its' just bad data. I downloaded a different set of places from: http://www.geonames.org/export/ and it appears to be lining up nicely.

**Comment by birdsarah** [comment]
also - google maps performs REALLY nicely on webgl :D

**Comment by bryevdv** [comment]
ping @hafen please check this out if you get a chance

**Comment by birdsarah** [comment]
In uploading new data for world cities in #2858, you can see the improvements with the new projected data:
before:
screen shot 2015-09-15 at 9 15 18 am

after:
screen shot 2015-09-15 at 9 17 38 am

**Comment by bryevdv** [comment]
When I run e.g. `maps.py` I get pretty odd axis labels:

screen shot 2015-09-15 at 5 56 06 pm

**Comment by birdsarah** [comment]
That is the coordinates of your data in web-mercator meters, which is what it's being displayed in on top of the google map. I could add some docs to explain.

**Comment by bryevdv** [comment]
Right I figured it was something along those lines, I just don't think that is very useful. If there is no way to improve the labeling of the axes we should leave off the axes in the examples. A little more tolerable since I don't think most maps have lat/lon labels.

**Comment by bryevdv** [comment]
Longer term there are a few possible solutions but most likely we will want a new ticker (that picks "nice" coordinates based on lat/lon derived from mercator meters) and a new formatter (that converts mercator meters before printing)

**Comment by birdsarah** [comment]
@bryevdv - I inadvertently pushed the cleanup of those examples to #2858 - they do stand alone so can be merged seperately with no worries.

Completely agree on those improvements - have started a new issue with those and some other thoughts I was having #2861

**Comment by bryevdv** [comment]
LGTM will merge both.

File: bokehjs/src/coffee/renderer/glyph/glyph.coffee [hunk 2]
```diff
@@ -71,11 +72,35 @@ class GlyphView extends ContinuumView
...
     # set all the coordinate fields
     for name, prop of @coords
       @[name] = prop.array(source)
 
+    if @renderer.plot_model.use_map
+      if @x?
+        [@x, @y] = @project_xy(@x, @y)
+      if @xs?
+        [@xs, @ys] = @project_xsys(@xs, @ys)
+
```
Comment (bryevdv) [inline bokehjs/src/coffee/renderer/glyph/glyph.coffee]: We should figure out how to generalize all this a bit better, but that can be for later.

Comment (birdsarah) [inline bokehjs/src/coffee/renderer/glyph/glyph.coffee]: Definitely! And probably put it in a set of geo utils. Also see related ideas in #2861.

**Comment by hafen** [comment]
Sorry I'm a bit late here - I will take a look at this. Thanks so much for this!  Very excited to try it out.

**Comment by hafen** [comment]
I gave it a shot and here are my results.  This was building from master.

Before:

before

After:

after

There is a change, but not correct.  I could be doing something wrong.

**Comment by birdsarah** [comment]
@hafen. I had something like this happen. I thought that there was a problem with the map_cities dataset. See: https://github.com/bokeh/bokeh/pull/2858

However, some days later I came back to it and it all magically worked fine.

Can I recommend testing again with complete clean out of all build directories, site packages, node_modules etc etc, some crazy hard refresh etc. 
and/or 
can you send me the exact dataset you're using.

**Comment by hafen** [comment]
@birdsarah interesting.  I cleaned things out (didn't clean out global node modules) and it still looks the same.  Here's the data: https://gist.github.com/hafen/0df3df084d03a2f4136f.  Perhaps I could get your js/css files and try those?

**Comment by hafen** [comment]
Whoa, that was strange - right after posting my message above, I went back to the page of my plot and everything was in the right place!  I don't know what happened.  But then when I refreshed the page, it went back to the offset.

**Pull request merged**